packages/rpc/src/vatidation/TracerResultParser.ts

```diff
@@ -0,0 +1,756 @@
+                if (associatedWith(slot, sender, entitySlots)) {
+                    if (userOp.initCode.length > 2) {
+                        // special case: account.validateUserOp is allowed to use assoc storage if factory is staked.
+                        // [STO-022], [STO-021]
+                        if (
+                            !(
+                                entityAddr === sender &&
+                                isStaked(stakeInfoEntities.factory)
+                            )
+                        ) {
+                            requireStakeSlot = slot
+                        }
+                    }
+                } else if (associatedWith(slot, entityAddr, entitySlots)) {
+                    // [STO-032]
+                    // accessing a slot associated with entityAddr (e.g. token.balanceOf(paymaster)
+                    requireStakeSlot = slot
+                } else if (addr === entityAddr) {
+                    // [STO-031]
+                    // accessing storage member of entity itself requires stake.
+                    requireStakeSlot = slot
+                } else if (writes[slot] === undefined) {
+                    // [STO-033]: staked entity have read-only access to any storage in non-entity contract.
+                    requireStakeSlot = slot
+                } else {
+                    // accessing arbitrary storage of another contract is not allowed
+                    const readWrite = Object.keys(writes).includes(addr)
+                        ? "write to"
+                        : "read from"
+
+                    const message = `${entityTitle} has forbidden ${readWrite} ${nameAddr(
+                        addr,
+                        entityTitle
+                    )} slot ${slot}`
+
+                    throw new RpcError(
+                        message,
+                        ValidationErrors.OpcodeValidation,
+                        {
+                            [entityTitle]: entStakes?.addr
+                        }
+                    )
+                }
+            }
+
+            // if addr is current account/paymaster/factory, then return that title
+            // otherwise, return addr as-is
+            function nameAddr(addr: string, currentEntity: string): string {
+                const [title] =
+                    Object.entries(stakeInfoEntities).find(
+                        ([title, info]) =>
+                            info?.addr.toLowerCase() === addr.toLowerCase()
+                    ) ?? []
+
+                return title ?? addr
+            }
+
+            requireCondAndStake(
+                requireStakeSlot !== undefined,
+                entStakes,
+                `unstaked ${entityTitle} accessed ${nameAddr(
+                    addr,
+                    entityTitle
+                )} slot ${requireStakeSlot}`
+            )
+        }
+
+        // [EREP-050]
+        if (entityTitle === "paymaster") {
```

Comment:
should we not be using an enum for these?

Reply:
this is typed so it will give a type error if anything else is used.